### PR TITLE
Visitors and other utility traits

### PR DESCRIPTION
**Pull request by queerviolet**
This PR adds the visitor pattern to the query and schema ASTs. It also adds a `Name` trait, which AST nodes implement if they have a `name`. Other similarly-useful traits will be forthcoming.

In addition to being able to visit nodes, we're also adding the ability to map over the AST tree. Tree maps are like visitors, only they return an arbitrary output type.

### CODE REVIEW

**Comment by jbaxleyiii** [comment]
@queerviolet since this is a pretty large PR (which looks to be a lot of snapshots!) can you help me understand the best way to go about a review of it? Is there a set of files I should be looking at in any order, a design / usage goal, etc? I'm really familiar with the visitor in graphql-js so I suspect I can blind review this but I'd love some general direction!

**Comment by queerviolet** [comment]
Yes! There's [usage description in `lib.rs`](https://github.com/apollographql/rust/pull/83/files#diff-68d4d6986a06726fa25916ce744c6504) and the `visit.rs` and `name.rs` files (in `graphql-parser` and `graphql-parser/schema` and `graphql-parser/query`) contain the real changes.

The question we should be asking is: can we use this for query planning? And I think we can. In particular, I believe that building a query plan can be expressed as a *map* operation. The map functionality introduced here uses a visitor which maintains a stack and implements a `merge` method describing how child output nodes are merged into their parents. Additionally, the `Map` is called mutably, and so can track additional state. I believe that will cover our needs, but welcome additional thoughts on it.

There *are* a lot of snapshot changes, mostly because I changed the snapshots to include the source so they're easier to review individually at a glance. I've looked through them to make sure they're okay, but some spot checking there would be great.

File: graphql-parser/src/lib.rs [hunk 1]
```diff
@@ -89,6 +89,96 @@
...
+//! let tx = ast.map(ToIndentedNodeTypes{});
+//! pretty_assertions::assert_eq!(tx.output, Some(String::from("query
+//!   query_def
+//!     sel_set
+//!       sel
+//!         sel_set
+//!       sel
+//!         sel_set
+//!           sel")));
+//!# Ok::<(), query::ParseError>(())
```
[the remark this answers is not in the document]
Comment (queerviolet) [reply graphql-parser/src/lib.rs:180]: It tells the test runner what the return type of the test is.

File: graphql-parser/src/visit.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    ///
+    /// Implementing this method lets you update and/or replace parent nodes
+    /// by including data from their children. The default implementation returns
+    /// the parent output node unchanged.
+    fn merge(&mut self, parent: Self::Output, child: &Self::Output) -> Self::Output {
+        parent
+    }
+}
+
+/// The output of a call to `map` is a Mappping
```
[the remark this answers is not in the document]
Comment (queerviolet) [reply graphql-parser/src/visit.rs]: Commented on the other PR, but basically `map` is mutable, and we're going to aggregate stuff into it, so we'd like to get it back.

We could return (output, map) or something, but it doesn't seem that useful to exclude the empty stack vec?

[the remark this answers is not in the document]
Comment (Enrico2) [reply graphql-parser/src/visit.rs]: The `map` function doesn't take a mutable `map: M`, which led me to this comment, but I get what you're saying here, worst case we revisit if we only end up using `.output` from the return type.

File: graphql-parser/src/schema/visit.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    fn schema_def<'a>(&mut self, def: &Definition<'a>, stack: &[Self::Output]) -> Self::Output;
+    fn field<'a>(&mut self, field: &Field<'a>, stack: &[Self::Output]) -> Self::Output;
+    fn input_value<'a>(
+        &mut self,
+        input_value: &InputValue<'a>,
+        stack: &[Self::Output],
+    ) -> Self::Output;
+}
+
+impl<M: Map> Visitor for visit::Fold<M> {
```
[the remark this answers is not in the document]
Comment (queerviolet) [reply graphql-parser/src/schema/visit.rs:29]: done

File: graphql-parser/src/lib.rs [hunk 4]
```diff
@@ -127,7 +127,7 @@
 //! let mut fields = Fields { output: vec![] };
 //! ast.accept(&mut fields);
 //! assert_eq!(fields.output, vec!["fieldA", "fieldB", "fieldC"]);
-//!# Ok::<(), schema::ParseError>(())
+//!# Ok::<(), graphql_parser::ParseError>(())
```
Comment (Enrico2) [inline graphql-parser/src/lib.rs:130]: Hm so I think these lines are not required? I found out that the `#` thing only hides that line from the docstring, but still runs that code. In this case (and in line 180) we don't have a `fn` that returns `Result<(), ParseError>` like in line 49, so this code is run as if it's part of a `main` (I'm pretty sure), which means it has no return type.

File: graphql-parser/src/query/visit.rs [hunk 5]
```diff
@@ -104,7 +103,8 @@ impl<'a> Node for Selection<'a> {
 }
 
 mod tests {
```
Comment (Enrico2) [inline graphql-parser/src/query/visit.rs:105]:
```suggestion
#[cfg(test)]
mod tests {
```

**Comment by Enrico2** [comment]
(Remember to `cargo fmt` before pushing, the build fails if there are fmt changes)

File: graphql-parser/src/lib.rs (hunk 4 above)
Comment (queerviolet) [reply graphql-parser/src/lib.rs:130]: https://doc.rust-lang.org/rustdoc/documentation-tests.html#using--in-doc-tests

File: graphql-parser/src/visit.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (queerviolet) [reply graphql-parser/src/visit.rs]: it takes ownership, and the methods on `schema::Map` and `query::Map` take `&mut self`

**Comment by Enrico2** [comment]
(Approved pending build passing)

**Pull request merged**